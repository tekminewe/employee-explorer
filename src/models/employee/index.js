import axios from 'axios';

export const endpoint = 'http://api.additivasia.io/api/v1/assignment/employees';

const getSubordinate = async (count, subordinateSet) => {
  const employeeName = [...subordinateSet][count];
  const response = await axios.get(`${endpoint}/${employeeName}`);
  const { data } = response;
  const [, subordinate] = data;
  if (subordinate) {
    const subordinates = subordinate['direct-subordinates'];
    subordinates.forEach((sub) => subordinateSet.add(sub));
  }

  if (count === subordinateSet.size - 1) {
    return subordinateSet;
  } else {
    return await getSubordinate(++count, subordinateSet);
  }
}

export const getSubordinateForEmployee = async (employeeName) => {
  const subordinateSet = new Set();
  subordinateSet.add(employeeName);

  try {
    const set = await getSubordinate(0, subordinateSet);
    set.delete(employeeName);
    return [...set];
  } catch (error) {
    throw new Error('Unexpected error occured');
  }
};
