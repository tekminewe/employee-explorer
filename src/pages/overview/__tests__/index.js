import React from 'react';
import { render, act } from '@testing-library/react';
import { getSubordinateForEmployee } from '../../../models/employee';

import OverviewPage from '../index';

jest.mock('../../../models/employee');

describe('OverviewPage', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('should get employee\'s subordinates', async () => {
    const props = {
      match: {
        params: {
          id: 'Henry Lee',
        },
      },
    }

    getSubordinateForEmployee.mockResolvedValue([
      'Subordinate 1',
      'Subordinate 2',
      'Subordinate 3',
    ]);

    let component;
    await act(async () => {
      component = render(<OverviewPage {...props} />);
    })
    expect(component.container).toMatchSnapshot();
  });

  test('should show loading when getting subordinates', async () => {
    const props = {
      match: {
        params: {
          id: 'Henry Lee',
        },
      },
    }

    getSubordinateForEmployee.mockResolvedValue([
      'Subordinate 1',
      'Subordinate 2',
      'Subordinate 3',
    ]);

    await act(async () => {
      const { container } = render(<OverviewPage {...props} />);
      expect(container).toMatchSnapshot();
    })
  });

  test('should show general error message if other error', async () => {
    const props = {
      match: {
        params: {
          id: 'Henry Lee',
        },
      },
    };

    getSubordinateForEmployee.mockRejectedValue(new Error('Other error'));

    let component;
    await act(async () => {
      component = render(<OverviewPage {...props} />);
    })
    expect(component.container).toMatchSnapshot();
  });
});
