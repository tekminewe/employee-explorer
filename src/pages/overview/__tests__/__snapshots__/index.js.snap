// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewPage should get employee's subordinates 1`] = `
<div>
  <div
    class="sc-AykKC jjVxQu"
  >
    <div
      class="sc-AykKD kOsfAI"
    >
      <h1
        class="sc-AykKE cuheMn"
      >
        Employee Overview
      </h1>
    </div>
    <div
      class="sc-AykKF hBaFEr"
    >
      <p>
        Subordinates of employee 
        <span
          class="sc-AykKG bYBuez"
        >
          Henry Lee
        </span>
      </p>
      <ul
        class="sc-AykKH klLnVA"
      >
        <li
          class="sc-AykKI fVbxAQ"
        >
          Subordinate 1
        </li>
        <li
          class="sc-AykKI fVbxAQ"
        >
          Subordinate 2
        </li>
        <li
          class="sc-AykKI fVbxAQ"
        >
          Subordinate 3
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`OverviewPage should show general error message if other error 1`] = `
<div>
  <div
    class="sc-AykKJ dWHlJx"
  >
    <svg
      class="sc-fzXfLO gjphZj"
    >
      alarm.svg
    </svg>
    <h1
      class="sc-AykKK cEQxlF"
    >
      Opps something goes wrong
    </h1>
  </div>
</div>
`;

exports[`OverviewPage should show loading when getting subordinates 1`] = `
<div>
  <div
    class="sc-AykKC jjVxQu"
  >
    <div
      class="sc-AykKD kOsfAI"
    >
      <h1
        class="sc-AykKE cuheMn"
      >
        Employee Overview
      </h1>
    </div>
    <div
      class="sc-AykKF hBaFEr"
    >
      <div
        class="sc-fzXfLP fOEUNm"
        data-testid="loading"
      />
    </div>
  </div>
</div>
`;
